src/priceClient/priceClient.ts

```diff
@@ -0,0 +1,212 @@
+import assert from "assert";
+import { CoinGeckoPrice } from "../coingecko";
+import { Logger } from "../relayFeeCalculator"; // @todo: Relocate Logger to utils?
+
+export { Logger } from "../relayFeeCalculator"; // Permit adapters to import a common definition.
+export type TokenPrice = CoinGeckoPrice; // Temporary inversion; CoinGecko should source from here.
+
+// Represents valid source for spot prices (Across API, CoinGecko, ...)
+export interface PriceFeedAdapter {
+  readonly name: string;
+  getTokenPrice(address: string, currency: string, platform: string): Promise<TokenPrice>;
+  getTokenPrices(addresses: string[], currency: string, platform: string): Promise<TokenPrice[]>;
+}
+
+export type PriceCache = {
+  [address: string]: TokenPrice;
+};
+
+export function msToS(ms: number): number {
+  return Math.floor(ms / 1000);
+}
+
+export class PriceClient implements PriceFeedAdapter {
+  private static instance: PriceClient | undefined;
+  public readonly name: string = "PriceClient";
+  private _maxPriceAge = 300; // seconds
+  private priceFeeds: PriceFeedAdapter[] = [];
+  private prices: {
+    [platform: string]: {
+      [currency: string]: PriceCache;
+    };
+  } = {};
+
+  protected constructor(protected logger: Logger) {}
+
+  public static get(logger: Logger): PriceClient {
+    if (!this.instance) this.instance = new PriceClient(logger);
+
+    return this.instance;
+  }
+
+  get maxPriceAge(): number {
+    return this._maxPriceAge;
+  }
+
+  set maxPriceAge(age: number) {
+    assert(age >= 0);
+    this.logger.debug({
+      at: "PriceClient#maxPriceAge",
+      message: `Setting maxPriceAge (S) ${this._maxPriceAge} => ${age}.`,
+    });
+    this._maxPriceAge = age;
+  }
+
+  // @todo: Could be nice to specify a priority rather than relying on order of addition.
+  addPriceFeed(priceFeed: PriceFeedAdapter): void {
```

Comment:
i think this dynamic adding of price feeds makes this code more complicated than it needs to be, does this need to be dynamic?  the alternative could be to just provide an array of price feeds in the constructor

Reply:
I think it could be done in the constructor, but since the constructor follows the singleton pattern initially established in the coingecko class, I'd be a little concerned about the impact that might have on the e2e test.

In the e2e test it's quite easy to throw away all existing price feeds in the beforeEach() callback, and then add new price feeds that are relevant to each test being executed. IMO that's quite nice and keeps everything fairly concise, but it's surely not the only way that it could be handled. Would you have an alternative suggestion?

Reply:
personally id prefer simpler business logic to simpler tests, but in the case of testing you can have separate describe blocks for each configuration of the price feed, rather than combining them all into a single describe block. it would be about equivalent to what you have

Reply:
I don't really see the runtime addition of price feeds as being overly complex (the methods are quite small), but I concede that this is a feature that would probably only be used in tests. I looked at adding separate describe() blocks for each instance here, but it would have required a lot of additional boilerplate code. So I have instead implemented your change and have migrated away from the singleton pattern, since it doesn't seem right that the caller of get() will supply an array of price feeds but has no idea whether or not they will actually be used.

IMO this change does fortunately make the code simpler and I'm happy with the result, since I don't think we would actually need the singleton pattern in our usage of this.